### PR TITLE
Give the default failure "reason" when commands failed

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes
Without giving the default reason, mappers will parse an error when command status is failed but no "reason" is given. However, "reason" filed should be optional in command payloads.

This bug was found during writing a unit test for firmware_update support.  
```
[te/device/main///cmd/firmware_update/c8y-mapper-2023-11-16T14:02:31.419901262Z] {
   "status":"failed",
   "remoteUrl":"https://speed.hetzner.de/10GB.bin",
   "name":"10GB",
   "version":"2.0.0"
}
[te/errors] missing field `reason` at line 6 column 1
```

The bahaviour is covered by the unit test `handle_log_upload_executing_and_failed_cmd_for_main_device()`.

Maybe worth adding a robot test?

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_api/src/messages.rs [hunk 1]
```diff
@@ -489,10 +489,15 @@ pub enum CommandStatus {
     Executing,
     Successful,
     Failed {
+        #[serde(default = "default_failure_reason")]
         reason: String,
     },
 }
 
+fn default_failure_reason() -> String {
+    "No failure reason is given".to_string()
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/messages.rs:498]: I'm not happy with this message which raises the question *given by who*.

That said I agree, it's hard to come with something sensible. "Unknown error"? In any case, as a user, this is frustrating.

Comment (didier-wenzek) [reply crates/core/tedge_api/src/messages.rs:498]: A default is indeed required.

Comment (albinsuresh) [reply crates/core/tedge_api/src/messages.rs:498]: Making the `reason` was indirectly solving that problem, right? It forces the users to provide a reason when there's a failure (as long as it is mandatory only for the `failed` state). When there's a failure, the user has to create a JSON message with the `status` payload anyway. So, why not make him add the `reason` field as well? Why give an option to the user to shoot himself in the foot?

Comment (didier-wenzek) [reply crates/core/tedge_api/src/messages.rs:498]: > So, why not make him add the reason field as well? Why give an option to the user to shoot himself in the foot?

The issue highlighted by @rina23q is that if by mistake no reason is given then the final error message is more then puzzling ( "missing field reason at line 6 column 1"). A default "Unknown reason" string will be clearer both for the agent developer and the end-user.

```suggestion
    "Unknown reason".to_string()
```

Comment (rina23q) [reply crates/core/tedge_api/src/messages.rs:498]: > So, why not make him add the reason field as well?
I believe that `reason` should not be mandatory. Highly recommended to add the reason for the failure though.

**Review by didier-wenzek — APPROVED** [review]
Approved.

**Pull request merged**